zjit/src/hir.rs

```diff
@@ -3561,23 +3561,248 @@ impl Function {
+                    | Insn::IfTrue { ref target, .. }
+                    | Insn::IfFalse { ref target, .. } => {
+                        successors.push(target.target.0 as u64);
+                    }
+                    _ => {}
+                }
+            }
+
+            // Store successors for this block.
+            block_successors.insert(block_id, successors.clone());
+
+            // Update predecessors for successor blocks.
+            for &succ_id in &successors {
+                block_predecessors
+                    .entry(BlockId(succ_id as usize))
+                    .or_default()
+                    .push(block_id.0 as u64);
+            }
+        }
+
+        (block_successors, block_predecessors)
+    }
+
+    /// Helper function to make an Iongraph JSON "instruction".
+    /// `uses`, `memInputs` and `attributes` are left empty for now, but may be populated
+    /// in the future.
+    fn make_iongraph_instr(ptr: u64, id: usize, inputs: Vec<Json>, opcode: &str, ty: &str) -> Json {
+        Json::object()
+            .insert("ptr", ptr)
+            .insert("id", id)
+            .insert("opcode", opcode)
+            .insert("attributes", Json::empty_array())
+            .insert("inputs", Json::Array(inputs))
+            .insert("uses", Json::empty_array())
+            .insert("memInputs", Json::empty_array())
+            .insert("type", ty)
+            .build()
+    }
+
+    /// Helper function to make an Iongraph JSON "block".
+    /// `attributes` is specific to Spidermonkey, so it is left empty..
+    fn make_iongraph_block(ptr: u64, id: u64, predecessors: Vec<u64>, successors: Vec<u64>, instructions: Vec<Json>) -> Json {
+        Json::object()
+            .insert("ptr", ptr)
+            .insert("id", id)
+            .insert("loopDepth", 0)
+            .insert("attributes", Json::empty_array())
+            .insert("predecessors", Json::array(predecessors))
+            .insert("successors", Json::array(successors))
+            .insert("instructions", Json::array(instructions))
+            .build()
+    }
+
+    /// Helper function to make an Iongraph JSON "function".
+    /// Note that `lir` is unpopulated right now as ZJIT doesn't use its functionality.
+    fn make_iongraph_function(pass_name: &str, mir_blocks: Vec<Json>) -> Json {
+        Json::object()
+            .insert("name", pass_name)
+            .insert("mir", Json::object()
+                .insert("blocks", Json::array(mir_blocks))
+                .build()
+            )
+            .insert("lir", Json::object()
+                .insert("blocks", Json::empty_array())
+                .build()
+            )
+            .build()
+    }
+
+    /// Generate an iongraph JSON pass representation for this function.
+    pub fn to_iongraph_pass(&self, pass_name: &str) -> Json {
+        let mut ptr_map = PtrPrintMap::identity();
+        ptr_map.map_ptrs = true;
+
+        let mut mir_blocks = Vec::new();
```

Comment:
let's use `hir` everywhere except for the literal strings that the tool expects to be `mir`